### PR TITLE
Add Windows Virtual Camera

### PR DESCRIPTION
**Pull request by jp9000**
### Description

The virtual camera adds the ability to use the output of OBS itself as a camera that can be selected within other Windows applications.  This is very loosely based upon the catxfish virtual camera plugin design, but greatly simplified/optimized.

### Motivation and Context

Motivation is to create a simplified/optimized virtual camera implementation for Windows based upon libdshowcapture rather than the 20 year old CBaseFilter (streams.h) design.

There is a shared memory queue, but instead of having 10-20 frames in the queue, there are now only 3 frames in the queue to minimize latency and reduce memory usage.

The frames of the shared memory queue are NV12 instead of YUYV, which reduces the memory and data copied, as well as eliminate unnecessary conversion from NV12.  Some programs (such as web browsers, which use webrtc to capture) do not support NV12 however, so an I420 conversion is provided, which is very cheap conversion in comparison.

The virtual camera filter itself is based upon the output filter within the libdshowcapture library, which was originally implemented for other purposes.  This is more ideal than the Microsoft example code because for one, it's far less convoluted, two, allows us to be able to customize the filter to our needs a bit more easily, and three, has much better RAII.  The Microsoft CBaseFilter/etc code comprised of about 30 source files and is twenty years old, where as the output filter comprises of two or three required source files which we already had.

Scaling is avoided whenever possible to minimize CPU usage, but if scaling must happen it uses a simple nearest neighbor implementation rather than depend on a third-party library for scaling.  The filter will remember the last OBS resolution if OBS is not opened first in order to minimize the possibility of scaling.  When used with OBS, no scaling will occur because OBS supports dynamic resolution changes from filters.

This addresses the Windows implementation of:
https://github.com/obsproject/rfcs/blob/master/accepted/0015-virtual-camera-support.md

This was extremely painful to implement and I hope I never have to work with DirectShow stuff ever again.

### How Has This Been Tested?

It's been tested with various programs that are commonly the target of a virtual camera, such as Zoom, Chrome, Microsoft Meetings, etc.  It's also been tested with OBS itself.

### Types of changes

- New feature (non-breaking change which adds functionality)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.

### CODE REVIEW

**Comment by WizardCM** [comment]
I wonder, would it be worth having a separate virtual camera DLL in ProgramData so it can be used by multiple installed/portable instances of OBS?

**Comment by jp9000** [comment]
I don't think it's necessary like it is with the vulkan registry entry.  Only one directory can be used at a time with the virtual camera's required registry entry.

**Comment by dodgepong** [comment]
A few comments regarding RFC compliance:

* After a lot of internal discussion, we've decided to omit the option to flip the camera horizontally, as this is virtually never needed, and will almost always end up in the user shooting themselves in the foot. The problem is that most video conferencing software shows the user's own view back to them mirrored, so users think that the OBS output is mirrored for some unknown reason. However, those apps will send the camera feed un-mirrored to meeting participants, so a Flip Horizontal button will let users will see their own preview un-mirrored but then make the actual view for other participants mirrored.
    * However, **it would behoove us to add some sort of warning/explanation somewhere in the UI about this behavior** so users will understand why their preview is mirrored, and won't come to us asking us for a flip horizontal option that they don't actually need.
* This PR appears to be missing the `--startvirtualcamera` command line option
* I'm not really a fan of the current placeholder image, and it doesn't conform to the RFC's recommendation to not require localization. My preference would be something more like this: https://user-images.githubusercontent.com/838450/85190398-2091a880-b286-11ea-9e62-6eb1b1bd1bb3.png
* This should include an update to the auto-config tool with a new primary use case option for outputting to a virtual camera

**Comment by exeldro** [comment]
The obs-virtualcam-module dlls get registered by the installer, but I can't find it registering in the updater.

**Comment by jp9000** [comment]
I added auto-updater code to register the modules, updated the image to dodgepong's suggestion, added the --startvritualcam command line parameter.  However I'm unsure at how the auto-config would work.  It'd basically just be "Hey I want to just use virtualcam", and it'd just say "okay you're good to go".  I can do that if that's fine with you.

**Comment by dodgepong** [comment]
The auto-config can just set the video output to 1080p30 and be done with it, I think. That is what the RFC recommends.

**Comment by WizardCM** [comment]
The "Start Virtual Camera" button misbehaves if you have no sources in your scene.

* Clicking "Yes" will make the popup appear a second time
* Clicking "No" will close the popup, but the button will remain checked (lit up) if you click it on the first popup

I'm not sure how I feel about generating a `obs-virtualcam.txt` in `%appdata%`, without a subdirectory.

Other than that, the camera works quite well. 

Was able to load it correctly in the following apps:
* Discord
* MS Teams
* Slack
* Skype UWP
* Skype Desktop
* Chromium-based Microsoft Edge
* OBS Studio 😉

It did not appear in the following apps:
* Windows 10's Camera app (though this is the case for most virtual cameras. only Snap Camera and FaceRig appear, could be a signature or whitelist thing)
* Snap Camera (similar to Win10's app, only FaceRig and NewTek NDI Video appear)

Other apps had weird behaviours:
* It appears within the list in FaceRig, however instead of loading the virtual output, it loads my C930e instead. It may be a bug in FaceRig, as the same happens with the LGP Stream Engine, however Snap Camera, NDI Video, OBS-Camera, and EOS Webcam Utility Beta behave correctly in FaceRig..
* TeamViewer seems to only show a single frame from when TeamViewer tries to load the camera. Toggling the output in OBS does not update the output in TeamViewer (whereas the VirtualCam plugin by catfish plays fine, but crashes TeamViewer if the output is not active)
* While it appears as an option in NDI Scan Converter, it never appears as an NDI input either in obs-ndi nor NDI Studio Monitor, as if it never initialises.

I will also test in Zoom tomorrow.

**Comment by dodgepong** [comment]
I admit that I'm not entirely sure where to put the notification to users about flipped webcams. Anybody have a good idea of how to inform users about that?

**Comment by Fenrirthviti** [comment]
> I admit that I'm not entirely sure where to put the notification to users about flipped webcams. Anybody have a good idea of how to inform users about that?

I saw we don't. It's not really our responsibility to inform users how other applications might work. The conferencing applications should be communicating this, and they have already accepted responsibility for it in many cases (see: the issue regarding Teams where they have admitted they communicate this poorly and will be updating to make it less confusing). This will probably be a common question through support, but the more I think about it the more I think we're stressing out over nothing.

**Comment by dodgepong** [comment]
It may not be our responsibility, strictly speaking, but that won't prevent us from having an absolute deluge of support requests about this. A warning would be a preemptive action to educate users about this detail so that we don't have to deal with requests about this multiple times a day on the forums and in support chat.

If we really want to YAGNI this then I suppose we'll find out when we release a beta/RC. At the very least, this detail deserves a mention in the release notes.

**Comment by Fenrirthviti** [comment]
Sure, my point maybe was missed with the last line there due to ambiguity on my end, apologies. While it might be a common question, I don't feel it will be as big of a deal as we think it is and it won't be overwhelming for support. The only thing I can think of in terms of how to notify is to make it a popup, and that's just bad UX and we already have too many of those as is. A proper user feedback system would make this easy, but lacking that, I don't think there is a good way to communicate it.

**Comment by dodgepong** [comment]
If the virtualcam had a dedicated section under settings, that would be the place to put it. However, in the interest of simplicity, we've so far opted not to include any options for the virtualcam.

...Are there any options we do want to add to this feature so that we would have a dedicated place to put these sorts of messages? Perhaps we should have an auto-start checkbox after all?

**Comment by jp9000** [comment]
Fixed the "no sources" bug you mentioned.

**Comment by fatbrain** [comment]
Tested this pull-request yesterday, found it strange that `vcam_installed(false)` was called with `b64` set to `false`, which in essence disabled the virtual-cam feature on my x64 build. Is there a reason for that?

**Comment by jp9000** [comment]
@fatbrain That does not enable/disable virtual camera, all that function does is test whether the virtual camera filter is installed.  I originally had it test both 32bit and 64bit, but I realized that some people might be running 32bit windows, so I just changed it to only test 32bit instead, seeing as if they had the 32bit filter installed, they would also have the 64bit filter installed.  The installer and auto-updater will install both.

To enable virtual camera for 64bit, you need to call regsvr32 on obs-virtualcam-module64.dll as well, not just obs-virtualcam-module32.dll.  That's probably your issue.

**Comment by ahmadnassri** [comment]
please pardon my limited understanding of how this works, and the fact that I pulled the build off azure and started using it despite it not being "released" yet :)

I have two GPUs, and I've configured OBS to use a specific GPU for streaming and recording ... it doesn't seem that the virtual camera is following those settings ... (and maybe it shouldn't by design?) .... when I "start" the the vcam I notice load on GPU 0 and nothing on GPU 1 ... whereas my recording settings are pointing to GPU 1 ...

I've always assumed the 3rd party Virtual Cam plugin followed the recording settings, and now I'm basing this assumption on this new plugin / feature as well ... am I wrong in assuming so?

- GPU 0 (embedded AMD Vega M) is my main display & monitors
- GPU 1 (NVIDIA GTX 980) is only used for stream encoding / RTX Voice app 

I'm not doing any game / high demand streaming, so while I know this is not an idea GPU setup, it's fine for my uses, which are 90% conference call with virtual cam + local recording & 10% streaming talks / coding videos.

File: plugins/win-dshow/virtualcam-module/virtualcam-module.cpp [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+	if (reg) {
+		ComPtr<IMoniker> moniker;
+		REGFILTER2 rf2;
+		rf2.dwVersion = 1;
+		rf2.dwMerit = MERIT_DO_NOT_USE;
+		rf2.cPins = 1;
+		rf2.rgPins = &AMSPinVideo;
+
+		hr = fm->RegisterFilter(CLSID_OBS_VirtualVideo,
+					L"OBS Video Output", &moniker,
```
Comment (dodgepong) [inline plugins/win-dshow/virtualcam-module/virtualcam-module.cpp:196]: This should be "OBS Virtual Camera" to be in line with the RFC.

**Comment by jaykijay** [comment]
> I wonder, would it be worth having a separate virtual camera DLL in ProgramData so it can be used by multiple installed/portable instances of OBS?

This would have been great!
Missing this already :(

**Comment by twoelfer** [comment]
As far as i can tell from your discussion, virtual cam is supposed to use the (previous) used resolution from obs. or the one currently configured. I have checked this using zoom (where obs virtual cam can be shared as a 'additional camera') and it works quite good.

However, it does not work in MS Teams: Resolution in Teams always looks like its 640x380 (or suchlike). Any ideas what might be causing this? Am i using the wrong build/version? Anything i can do to help with this issue?

**Comment by twoelfer** [comment]
So it appears: the above issue is only happening for the _recording_ of a meeting. This was how i tested - looked at the recoreded session, and the recording was very bad. (when recording a normal screen share, the quality is ok.). Today i tested with an actual person watching and it turns out: quality is ok. So this seems to be something having to do with Teams while recording, not something haveing todo with the virtual camera.

I validate this by installing the ndi tools and plugin and i found the same behaviour: Recording quality is very bad while the actual presentation looks ok for the other person.

**Pull request merged**